src/workerd/jsg/modules.h

```diff
@@ -409,25 +409,53 @@ class ModuleRegistryImpl final: public ModuleRegistry {
           if (type == Type::BUILTIN && entries.find(Key(path, Type::BUNDLE)) != kj::none) {
             continue;
           }
-
-          // The body of this callback is copied from `compileWasmGlobal` in src/workerd/server/workerd-api.c++.
-          addBuiltinModule(specifier, [specifier, module, this](Lock& lock) {
-            lock.setAllowEval(true);
-            KJ_DEFER(lock.setAllowEval(false));
-
-            // Allow Wasm compilation to spawn a background thread for tier-up, i.e. recompiling
-            // Wasm with optimizations in the background. Otherwise Wasm startup is way too slow.
-            // Until tier-up finishes, requests will be handled using Liftoff-generated code, which
-            // compiles fast but runs slower.
-            AllowV8BackgroundThreadsScope scope;
-            auto wasmModule = jsg::compileWasmModule(lock, module.getWasm().asBytes(), this->observer);
-            return jsg::ModuleRegistry::ModuleInfo(
-                  lock,
-                  specifier,
-                  kj::none,
-                  jsg::ModuleRegistry::WasmModuleInfo(lock, wasmModule));
-          }, type);
-          continue;
+          switch(module.which()) {
+          case Module::WASM:
+            // The body of this callback is copied from `compileWasmGlobal` in src/workerd/server/workerd-api.c++.
+            addBuiltinModule(specifier, [specifier, module, this](Lock& lock) {
+              lock.setAllowEval(true);
+              KJ_DEFER(lock.setAllowEval(false));
+
+              // Allow Wasm compilation to spawn a background thread for tier-up, i.e. recompiling
+              // Wasm with optimizations in the background. Otherwise Wasm startup is way too slow.
+              // Until tier-up finishes, requests will be handled using Liftoff-generated code, which
+              // compiles fast but runs slower.
+              AllowV8BackgroundThreadsScope scope;
+              auto wasmModule = jsg::compileWasmModule(lock, module.getWasm().asBytes(), this->observer);
+              return jsg::ModuleRegistry::ModuleInfo(
+                    lock,
+                    specifier,
+                    kj::none,
+                    jsg::ModuleRegistry::WasmModuleInfo(lock, wasmModule));
+            }, type);
+            continue;
+          case Module::DATA:
+            addBuiltinModule(specifier, [specifier, module](Lock& lock) {
+              auto value = kj::heapArray(module.getData().asBytes().asBytes());
+              v8::Local<v8::ArrayBuffer> data;
```

Comment:
No change necessary, but just FYI: This looks fine for now but a key thing to keep in mind with DATA modules is that they are exposed as *mutable* `ArrayBuffer` instances. In other words, if someone does `import { default as foo } from 'foo'` and `foo` is a `DATA` module, then I can `new Uint8Array(foo)` and modify the content of the underlying `ArrayBuffer`. If these built-ins are INTERNAL only modules then this really isn't a problem. If these are exposed to user code (i.e. BUILTIN type) then there's a non-zero risk of users modifying the buffer. Generally this isn't a problem since it would only impact the one Worker instance and this one copy of the DATA, but it's worth pointing out.

Reply:
I think we should eventually make these immutable buffers that can be shared between multiple instances to save on copying. If a mutable copy is needed, the consuming code can do `new Uint8Array(arrayBuff).slice()` so I think it would make sense to always make them immutable.

On the other hand, since you're already exposing user data modules as mutable copies by default, changing that would be an API break. It might be weird if the internal/builtin data modules behaved differently than the user ones. OTOH it would not be an API break.

Reply:
Right. I guess it's not allowed to have an immutable `Uint8Array`, so the only benefit of this is people for some reason import the `data` module but don't use it. Because if we want to use the `Blob` data we'll have to get the `arrayBuffer()` out of it which is a copy.